include/El/optimization/bfgs.hpp

```diff
@@ -0,0 +1,226 @@
+    {
+        T alpha = (alpha_low + alpha_high)/T(2.0);
+        x_j = x0;
+        Axpy(alpha, p, x_j);
+        T fval = f(x_j);
+        if(Abs(fval)<= -c2*f0)
+        {
+            return alpha;
+        }
+        gradient( x_j, g2);
+        auto fval_dash = Dot(p, g2);
+        if(  fval_dash*(alpha_high - alpha_low) >= 0)
+        {
+            alpha_high = alpha_low;
+        }
+        alpha_low = alpha;
+    }
+    return (alpha_high+alpha_low)/T(2.0);
+}
+/***
+ * This function approximately minimizes the function min_\alpha f(x0 + alpha*p)
+ * In some sense minimizing this function is "ideal" however, this may take a lot of work, and since
+ * we are using this inside of a larger algorithm on an approximate problem, there is no reason to believe
+ * that this is the "best" step. Instead we just hope for a step with "sufficient" decrease, and then we
+ * hope to prove that if {x_i} approaches the minimum that this function converges.
+ * This algorithm attempts to satsify the weak wolf conditions.
+ */
+template< typename T, typename Function, typename Gradient>
+T lineSearch( const Function& f, const Gradient& gradient,
+              const DistMatrix<T>& g, Int D,
+              const DistMatrix<T>& x0, const DistMatrix<T>& p,
+              Int maxIter=100, T c1=Pow(limits::Epsilon< Base<T> >(), Base<T>(1)/Base<T>(4)), T c2=T(9)/T(10))
+    {
+
+        if( c1 > c2) { std::swap(c1, c2); }
+
+        T f0 = f(x0);
+        DistMatrix<T> g2(D, 1, x0.Grid());
+        T f0_dash = Dot(p,g);
+        T  alpha(1);
+        T  alpha_prev(0);
+        T  alphaMax(1e3);
+        T fvalPrev(0);
+        T fval(0);
+        DistMatrix<T> x_candidate(x0);
+        for(std::size_t iter = 0; iter < maxIter; ++iter)
+        {
+            x_candidate = x0;
+            Axpy(alpha, p, x_candidate);
+            fval = f(x_candidate);
+            //The quantity on the LHS is less than f0
+            //So we have found a direction where the function value
+            //has not gone down sufficiently.
+            if ( fval > f0 + c1*alpha*f0_dash || (iter > 0 && fval > fvalPrev) )
+            {
+                return Zoom(f, gradient, f0, x0, p, alpha_prev, alpha, c1, c2);
+            }
+            gradient( x_candidate, g2);
+            auto fval_dash = Dot(p, g2);
+            if(  Abs(fval_dash) <= -c2*f0_dash)
+            {
+                return alpha;
+            }
+            if( fval_dash > 0)
+            {
+                return Zoom(f, gradient, f0, x0, p, alpha, alpha_prev, c1, c2);
+            }
+            alpha_prev = alpha;
+            fvalPrev = fval;
+            alpha = std::min(2*alpha, alphaMax);
+        }
+    return alpha;
+}
+
+namespace detail {
+/***
+ * This class stores the updates to the inverse hessian of F, and provides the ability to apply inverse hessian.
+ */
+template< typename T>
+struct HessianInverseOperator {
+
+    typedef DistMatrix<T> Matrix;
```

Comment:
*Please* do not typedef `El::DistMatrix` to `El::Matrix` when `El::Matrix` already exists...